spec/controllers/users/phone_confirmation_controller.rb

```diff
@@ -0,0 +1,106 @@
+require 'rails_helper'
+
+describe Users::PhoneConfirmationController, devise: true do
+  describe 'before_actions' do
+    it 'includes authentication checks' do
+      expect(subject).to have_filters(
+        :before,
+        :authenticate_user!,
+        :check_for_unconfirmed_mobile
+      )
+    end
+  end
+
+  describe '#new' do
+    before do
+      sign_in_as_user
+    end
+
+    context 'when :unconfirmed_mobile is not set in session' do
+      it 'redirects to rool_url' do
+        get :new
+
+        expect(response).to redirect_to(root_url)
+      end
+    end
+
+    context 'when :unconfirmed_mobile is set in session' do
+      before do
+        subject.user_session[:unconfirmed_mobile] = '+1 (555) 555-5555'
+      end
+
+      it 'generates a new confirmation code in the session' do
+        expect(subject.user_session[:phone_confirmation_code]).to be_nil
+        get :new
+        expect(subject.user_session[:phone_confirmation_code]).not_to be_nil
+      end
```

Comment:
Instead of checking for any non-nil value, I would make sure it only contains digits and has an amount of digits that corresponds to `Devise.direct_otp_length`. To ensure it's a random number, we can do something like this:
```ruby
expect(SecureRandom).to receive(:random_number).with(10**Devise.direct_otp_length)
```